src/main/cpp/src/mixed_sort_merge_join.cu

```diff
@@ -0,0 +1,759 @@
+ * limitations under the License.
+ */
+
+#include "mixed_sort_merge_join.hpp"
+
+#include <cudf/ast/detail/expression_evaluator.cuh>
+#include <cudf/ast/detail/expression_parser.hpp>
+#include <cudf/ast/expressions.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/join/sort_merge_join.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda_runtime_api.h>
+#include <thrust/copy.h>
+#include <thrust/count.h>
+#include <thrust/distance.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/remove.h>
+#include <thrust/sort.h>
+#include <thrust/tuple.h>
+#include <thrust/unique.h>
+
+namespace spark_rapids_jni {
+
+namespace {
+
+// =============================================================================
+// HELPER FUNCTIONS (Used by both conditional and non-conditional joins)
+// =============================================================================
+template <bool has_nulls>
+__global__ void filter_join_indices_kernel(
+  cudf::size_type const* __restrict__ left_indices,
+  cudf::size_type const* __restrict__ right_indices,
+  cudf::size_type num_pairs,
+  cudf::table_device_view left_table,
+  cudf::table_device_view right_table,
+  cudf::ast::detail::expression_device_view device_expression_data,
+  bool* __restrict__ keep_mask)
+{
+  extern __shared__ char raw_intermediate_storage[];
+  cudf::ast::detail::IntermediateDataType<has_nulls>* intermediate_storage =
+    reinterpret_cast<cudf::ast::detail::IntermediateDataType<has_nulls>*>(raw_intermediate_storage);
+  auto thread_intermediate_storage =
+    &intermediate_storage[threadIdx.x * device_expression_data.num_intermediates];
+
+  auto evaluator = cudf::ast::detail::expression_evaluator<has_nulls>(
+    left_table, right_table, device_expression_data);
+
+  cudf::thread_index_type const idx    = blockIdx.x * blockDim.x + threadIdx.x;
+  cudf::thread_index_type const stride = blockDim.x * gridDim.x;
+
+  for (cudf::thread_index_type i = idx; i < num_pairs; i += stride) {
+    auto output_dest                = cudf::ast::detail::value_expression_result<bool, has_nulls>();
+    cudf::size_type const left_idx  = left_indices[i];
+    cudf::size_type const right_idx = right_indices[i];
+
+    evaluator.evaluate(output_dest, left_idx, right_idx, 0, thread_intermediate_storage);
+
+    keep_mask[i] = output_dest.is_valid() && output_dest.value();
+  }
+}
+
+/**
+ * @brief Filter join results by evaluating conditional expression
+ */
+template <bool has_nulls>
+std::pair<std::unique_ptr<rmm::device_uvector<cudf::size_type>>,
+          std::unique_ptr<rmm::device_uvector<cudf::size_type>>>
```

Comment:
Style-wise these days you'd just normally return the `rmm::device_vectors` directly and let the compiler RVO take care of it. This AI is living in the 2010s.